craft_application/services/project.py

```diff
@@ -0,0 +1,349 @@
+        :param platform: The name of the target platform.
+        :param build_on: The host architecture the build happens on. Defaults to the
+            current architecture.
+        :returns: A Project model containing the project rendered as above.
+        """
+        platforms = self.get_platforms()
+        if platform not in platforms:
+            raise errors.InvalidPlatformError(platform, sorted(platforms.keys()))
+
+        project = copy.deepcopy(self._load_raw_project())
+
+        GrammarAwareProject.validate_grammar(project)
+        self._app_preprocess_project(
+            project, build_on=build_on, build_for=build_for, platform=platform
+        )
+        self._expand_environment(project, build_for=build_for)
+
+        # Process grammar.
+        if "parts" in project:
+            emit.debug(f"Processing grammar (on {build_on} for {build_for})")
+            project["parts"] = grammar.process_parts(
+                parts_yaml_data=project["parts"],
+                arch=build_on,
+                target_arch=str(build_for),
+            )
+        project_model = self._app.ProjectClass.from_yaml_data(
+            project, self.resolve_project_file_path()
+        )
+
+        if not project_model.adopt_info:
+            missing_fields: set[str] = set()
+            for field in self._app.mandatory_adoptable_fields:
+                if not getattr(project_model, field, None):
+                    missing_fields.add(field)
+            if missing_fields:
+                missing = ", ".join(repr(field) for field in sorted(missing_fields))
+                raise errors.CraftValidationError(
+                    f"'adopt-info' not set and required fields are missing: {missing}"
+                )
+
+        return project_model
+
+    def update_project_environment(self, info: craft_parts.ProjectInfo) -> None:
+        """Update a ProjectInfo's global environment."""
+        info.global_environment.update(
+            {
+                "CRAFT_PROJECT_VERSION": info.get_project_var("version", raw_read=True),
+            }
+        )
+
+    @property
+    def is_rendered(self) -> bool:
+        """Whether the project has already been rendered."""
+        return self._project_model is not None
+
+    def get(self) -> models.Project:
+        """Get the rendered project.
+
+        :returns: The project model.
+        :raises: RuntimeError if the project has not been rendered.
+        """
+        if not self._project_model:
+            raise RuntimeError("Project not rendered yet.")
+        return self._project_model
+
+    def render_once(
+        self,
+        *,
+        platform: str | None = None,
+        build_for: str | None = None,
+    ) -> models.Project:
+        """Render the project model for this run.
+
+        This should only be called by the Application for initial setup of the project.
+        Everything else should use :py:method:`get`.
```

Comment:
Should this be enforced in some way, or at least return immediately if called more than once?

Reply:
This is what the first two lines do - if the project model is set, `render_once()` will raise a RuntimeError. The note here is to clarify that behaviour alongside the `:raises:` note.